frontend/src/graph/ui/Chart.vue

```diff
@@ -27,22 +28,74 @@ export default {
       type: Boolean,
       required: true,
     },
+    'doRestore': {
+      type: Boolean,
+      required: true,
+    },
+    'scale': {
+      type: Number,
+      required: true,
+    },
   },
   computed: {},
   data() {
     return {
+      imageHeight: 0,
+      imageWidth: 0,
+      originImageWidth: 0,
+      originImageHeight: 0,
+      graphZoom: null,
+      svgSelection: null,
+      zoomScale: null,
     };
   },
   watch: {
     doDownload: function(val) {
+      let chartScope = this;
       if (this.doDownload) {
-        let svg = this.$refs.graphSvg;
-        saveSvgAsPng(svg, "graph.png", {scale: 1.0});
-        this.$emit('triggerDownload', false);
+        this.restoreImage(true);
+        chartScope.$emit('triggerDownload', false);
       }
     },
+    doRestore: function(val) {
+      this.restoreImage(false);
+    },
+    scale: function(val) {
+      let k = this.zoomScale(val);
+      let svg = d3.select('svg');
+      svg.call(this.graphZoom.transform, d3.zoomIdentity.scale(k));
+    },
   },
+  methods: {
+    restoreImage(thenDownload) {
```

Comment:
I feel we should divide this into two functions. 
restore Image and download images are not tightly related. But I understand if there is technical difficulties, we can leave them as it is for now. 

Let's leave some more comments to explain this edge case. Should be helpful later.